File: loop_omp_acc.c
#include <math.h>
#include <stdio.h>
#include <stdlib.h>
#include <omp.h>
double rand_double() { return ((double) rand())/((double) RAND_MAX); }
int main(int argc, char *argv[])
{
  double t0,t1;uint N,i,j;double *restrict a,*restrict r,*restrict e,x,sum_omp,sum_acc,sum_loc;long npair;
  N = 20000;
  a = malloc(N*sizeof *a); r = malloc(N*sizeof *r); e = malloc(N*sizeof *e);
   
  for (i=0;i<N;i++) {    
    a[i] = rand_double(); 
    r[i] = 0.0; e[i] = 0.0;
  }

  // OpenMP loop
  npair = 0;
  sum_omp = 0.0;
  t0 = omp_get_wtime(); //clock();
#pragma omp parallel for private(i,j) default(shared) reduction(+:npair,sum_omp)\
  schedule(static) 
  for (i=0;i<N;i++) {
    for (j=0;j<N;j++) {
      npair += 1;
      r[i] += exp(a[i] + a[j]);
      e[i] += log(a[i] + a[j]);
      sum_omp += r[i]*0.1 + e[i]*0.2;
    }
  }
  t1 = omp_get_wtime();
  printf("# OMP wall time = %6.4f %ld %f\n", 
	 (t1-t0), npair, sum_omp);


  // OpenACC loop
  for (i=0;i<N;i++) {    
    r[i] = 0.0; e[i] = 0.0;
  }
  npair = 0;
  sum_acc = 0.0;
  t0 = omp_get_wtime();
#pragma acc data copyin(a[0:N]) copy(r[0:N],e[0:N])
#pragma acc kernels 
#pragma acc loop  reduction(+:npair, sum_acc)  
  for (i=0;i<N;i++) {
    sum_loc = 0.0;
    for (j=0;j<N;j++) {
      npair += 1;
      r[i] += exp(a[i] + a[j]);
      e[i] += log(a[i] + a[j]);
      sum_loc += r[i]*0.1 + e[i]*0.2;
    }
    sum_acc += sum_loc;
  }
  t1 = omp_get_wtime();
  printf("# ACC wall time = %6.4f %ld %f\n", 
	 (t1-t0), npair, sum_acc);
  free(a); free(r); free(e);
  return 0;
}

File: loop_omp_acc.f90
program main
    use omp_lib
    implicit none
    integer :: n        ! size of the vector
    real*8,allocatable :: a(:), r(:), e(:)
    real*8:: x, sum_omp, sum_acc, sum_loc
    integer*8 :: i, j, ierr, npair
    real:: t0,t1

    n = 1000
    allocate(a(n), r(n), e(n), stat=ierr)
    do i=1, n
       call random_number(x)
       a(i) = x
    end do
    r(:) = 0.0d0
    e(:) = 0.0d0
    npair = 0
    sum_omp = 0.0d0
    t0 = omp_get_wtime()
    !$omp parallel do  default(shared) private(i,j) reduction(+:npair,sum_omp)&
    !$omp schedule(static) 
    do i=1, n
       do j=1,n
          npair = npair + 1
          r(i) = r(i)+dexp(a(i) + a(j))
          e(i) = e(i)+dlog(a(i) + a(j))
          sum_omp = sum_omp + r(i)*0.1d0 + e(i)*0.2d0
       end do
    enddo
    !$omp end parallel do
    ! 
    t1 = omp_get_wtime()
    print *, "OMP took", t1 -t0, npair,sum_omp
    !
    r(:) = 0.0d0
    e(:) = 0.0d0
    npair = 0
    sum_acc = 0.0d0
    t0 = omp_get_wtime()
    !$acc data copyin(a(1:n)) copy(r(1:n),e(1:n))
    !$acc kernels 
    !$acc loop reduction(+:npair, sum_acc)
    do i=1, n
       sum_loc = 0.0
       do j=1,n
          npair = npair + 1
          r(i) = r(i)+dexp(a(i) + a(j))
          e(i) = e(i)+dlog(a(i) + a(j))
          sum_loc = sum_loc + r(i)*0.1d0 + e(i)*0.2d0          
       end do
       sum_acc = sum_acc + sum_loc
    enddo
    !$acc end kernels
    !$acc end data
    ! 

    t1 = omp_get_wtime()
    print *, "ACC took", t1 -t0, npair,sum_acc
    deallocate(a,r,e,stat=ierr)
end program
